File: services/auth/app/services/user_service.py
"""E1-H1: gestión de usuarios y roles."""
from datetime import timedelta

from sqlalchemy import select
from sqlalchemy.orm import Session

from app.config import settings
from app.domain import EstadoUsuario, Rol, can_manage_users
from app.mailer import build_activation_link, send_activation_email
from app.models import User
from app.schemas import UserCreate
from app.security import generate_token, utcnow


class ForbiddenError(Exception):
    pass


class DuplicateEmailError(Exception):
    pass


class UserNotFoundError(Exception):
    pass


def list_users(db: Session) -> list[User]:
    return list(db.execute(select(User).order_by(User.created_at.desc())).scalars())


def create_user(db: Session, actor_role: Rol, data: UserCreate) -> tuple[User, str]:
    # Escenario E1-H1: solo Gerencia/Administrador puede crear usuarios.
    if not can_manage_users(actor_role):
        raise ForbiddenError(
            "Solo el rol Gerencia/Administrador puede crear o modificar usuarios."
        )

    existing = db.execute(select(User).where(User.email == data.email)).scalar_one_or_none()
    if existing:
        # Escenario 2: correo duplicado, no se crea registro.
        raise DuplicateEmailError("Ya existe un usuario con ese correo")

    token = generate_token()
    user = User(
        name=data.name,
        email=data.email,
        role=data.role,
        linea_negocio=data.linea_negocio,
        status=EstadoUsuario.ACTIVO,
        activation_token=token,
        activation_token_expires_at=utcnow() + timedelta(hours=settings.activation_token_ttl_hours),
    )
    db.add(user)
    db.commit()
    db.refresh(user)

    link = build_activation_link(token)
    send_activation_email(user.email, link)
    return user, link


def _set_status(db: Session, actor_role: Rol, user_id: str, status: EstadoUsuario) -> User:
    if not can_manage_users(actor_role):
        raise ForbiddenError(
            "Solo el rol Gerencia/Administrador puede crear o modificar usuarios."
        )

    user = db.get(User, user_id)
    if user is None:
        raise UserNotFoundError("Usuario no encontrado")

    # Escenario 3: desactivar/activar solo cambia el estado; el historial
    # y los demás datos permanecen intactos.
    user.status = status
    db.commit()
    db.refresh(user)
    return user


def deactivate_user(db: Session, actor_role: Rol, user_id: str) -> User:
    return _set_status(db, actor_role, user_id, EstadoUsuario.INACTIVO)


def activate_user(db: Session, actor_role: Rol, user_id: str) -> User:
    return _set_status(db, actor_role, user_id, EstadoUsuario.ACTIVO)

File: src/lib/userService.ts
import type { PrismaClient, User } from "@prisma/client";
import type { Rol } from "@/lib/domain";
import { canManageUsers } from "@/lib/permissions";
import { createUserSchema, type CreateUserInput } from "@/lib/validators";
import { generateActivationToken, activationTokenExpiry } from "@/lib/tokens";
import { buildActivationLink, sendActivationEmail } from "@/lib/mailer";

export class ForbiddenError extends Error {}
export class DuplicateEmailError extends Error {}
export class UserNotFoundError extends Error {}

type Actor = { role: Rol };

/**
 * Subconjunto de PrismaClient que necesita este servicio, para poder
 * inyectar un doble de prueba en los tests unitarios sin levantar SQLite.
 */
export type UserRepository = Pick<PrismaClient["user"], "findUnique" | "create" | "update" | "findMany">;

export async function listUsers(db: UserRepository): Promise<User[]> {
  return db.findMany({ orderBy: { createdAt: "desc" } });
}

// Escenario 1 (E1-H1): alta de usuario exitosa.
// Escenario 2 (E1-H1): correo duplicado.
export async function createUser(
  db: UserRepository,
  actor: Actor,
  input: CreateUserInput
): Promise<{ user: User; activationLink: string }> {
  if (!canManageUsers(actor.role)) {
    throw new ForbiddenError(
      "Solo el rol Gerencia/Administrador puede crear o modificar usuarios."
    );
  }

  const data = createUserSchema.parse(input);

  const existing = await db.findUnique({ where: { email: data.email } });
  if (existing) {
    throw new DuplicateEmailError("Ya existe un usuario con ese correo");
  }

  const activationToken = generateActivationToken();

  const user = await db.create({
    data: {
      name: data.name,
      email: data.email,
      role: data.role,
      lineaNegocio: data.lineaNegocio,
      status: "ACTIVO",
      activationToken,
      activationTokenExpiresAt: activationTokenExpiry(),
    },
  });

  const activationLink = buildActivationLink(activationToken);
  await sendActivationEmail(user.email, activationLink);

  return { user, activationLink };
}

async function setUserStatus(
  db: UserRepository,
  actor: Actor,
  userId: string,
  status: "ACTIVO" | "INACTIVO"
): Promise<User> {
  if (!canManageUsers(actor.role)) {
    throw new ForbiddenError(
      "Solo el rol Gerencia/Administrador puede crear o modificar usuarios."
    );
  }

  const existing = await db.findUnique({ where: { id: userId } });
  if (!existing) {
    throw new UserNotFoundError("Usuario no encontrado");
  }

  return db.update({ where: { id: userId }, data: { status } });
}

// Escenario 3 (E1-H1): desactivación de usuario. Los datos e historial
// permanecen intactos: solo cambia el campo `status`, nunca se borra el registro.
export function deactivateUser(db: UserRepository, actor: Actor, userId: string) {
  return setUserStatus(db, actor, userId, "INACTIVO");
}

export function activateUser(db: UserRepository, actor: Actor, userId: string) {
  return setUserStatus(db, actor, userId, "ACTIVO");
}
